main/src/cgeo/geocaching/storage/DataStore.java

```diff
@@ -3832,6 +3836,25 @@ private static void deleteOrphanedRecords() {
         database.delete(dbTableExtension, "_type=" + DBEXTENSION_INVALID.id, null);
     }
 
+    private static void deleteOrphanedUDC() {
+        final Set<String> orphanedUDC = new HashSet<>();
+        queryToColl(dbTableCaches,
+                new String[]{"geocode"},
+                "SUBSTR(geocode,1," + InternalConnector.PREFIX.length() + ") = '" + InternalConnector.PREFIX + "' AND geocode NOT IN (SELECT geocode FROM " + dbTableCachesLists + " WHERE SUBSTR(geocode,1," + InternalConnector.PREFIX.length() + ") = '" + InternalConnector.PREFIX + "')",
```

Comment:
Performance of query might improve if in the 'geocode not in' subquery you remove the 'where substr...' part completely. That way SQLite might be able to use an index on dbtablecacheslists.geocode. 

Also, depending on indexes and cleverness of SQLite, usage of LIKE with wildcard % at end might be more performant than SUBSTR. But that's debatable.

Reply:
My thinking behind that condition was that typically you have very low UDC to all cache types ratio. If you leave out that condition, the database will do a cross-product of all caches against all cache to list relations, whereas with that condition it will only cross the much lower number of UDC against the cache/list relations.

Example: Let's have 20,000 caches, 100 of them UDC, and 25.000 list/cache relations.
Without the condition, the database will match 20,000 caches with 25,000 list/cache-relations, using a cross-product of 500 million entries, whereas with that condition, it will cross only 100 against 25,000, resulting in only 2.5 million entries to be tested.

Regarding the LIKE vs. SUBSTR I'm unsure which one is better. Both should be able to use the index, depending on the query optimizer. LIKE is typically optimized for strings starting with non-wildcard, on the other hand it's performing case conversion. (Not sure if that's the case for SQLite as well.)

Do we have any reasonably sized database to test this in real life?

Reply:
Using SUBSTR might require an additional functionbased index on the table. However, there is no clear answer which of both is more performant. People are undecided here...

https://stackoverflow.com/questions/7434668/performance-of-sql-comparison-using-substring-vs-like-with-wildcard

Reply:
I did a "EXPLAIN QUERY PLAN SELECT ..." and (at least with SQLite on the PC) both queries return the same execution plan (and both use the available indexes):

```
1) SCAN cg_caches USING COVERING INDEX in_caches_geo
2) LIST SUBQUERY 1
3) SCAN cg_caches_lists USING COVERING INDEX in_lists_geo
```

Leaving out the `WHERE` condition in the subquery I get a shorter query plan (as expected):

```
1) SCAN cg_caches USING COVERING INDEX in_caches_geo
2) USING INDEX in_lists_geo FOR IN-OPERATOR
```

Still would be nice if we could get some large real life database to run those queries against it.

What about including this PR in upcoming beta version and wait for possible reactions and/or Play console data?

Reply:
One of our "power users" is maybe willing to share their DB with cgeo support via mail. Maybe @MagpieFourtyTwo, @DaTechniker or @gitilo? We also understand if you don't want to do this...totally fine as well. 

In case we want to just merge it: In this case I would like to ask the users mentioned above to report back whether cgeo's startup time feels significantly slower in the near future.

Reply:
I am with @fm-sys that the PR should be merged to see how the change behaves in the field.

Wrt creating large databases: this should be very easy by writing a small test helper routine which creates a large number of (fake) Geocache objects using the methods of DataStore. I did this already e.g. for the unit tests for the new Filter framework, check e.g. class GeocacheFilterTestUtils. I created only single geocaches there (and then test whether they are correctly SQL-filtered), but by using a loop this could also be used for creating large DBS 
(But please make sure such methods are not executes as part of the unit tests when checking in :-))

Reply:
> power users

Although daily in use, my DB is not exceptionally big (~24 k entries), and as I seldom delete entries (just adding e. g. recommendations, solved mysteries etc. and even found ones will just be moved to a diff. list), I expect a vacuum to have a rather "moderate" effect at best. Nevertheless, I will of course check performance before/after and if there is a significant change, you can of course get the "before DB" for deeper analysis.

Reply:
Just tried ``Delete orphaned files`` with latest Nightly (2022.09.19-NB-fb9d87e) - "Maintenance" took 2:15 minutes and vacuumed the DB from 631 to 599 MB (size of sqlite backup before and after).

Not that much, one would think ... nevertheless, it feels *way* more performant now. And even objectively measured it has gained a lot: After first start it always took a few seconds(!) until it was "ready for action" (i. e. until it accepted a new command like entering a list or open the map), and now it's instantly responsive. Was not even able yet to get this "milky" list or map icon (sort of app freeze) I always got before, when I tapped the lists or map icon. And even while changing between lists, those (like e. g. one with ~7.5k caches or history) which took this kind of commemorative second before, are now instantly displayed.

IOW: Great stuff, BIG success!  ;)

Reply:
Thanks @MagpieFourtyTwo for your feedback. What you have observed is the result of #13423, the addition to reindex and compress the database. 😃 
This PR here is about immediately removing deleted UDC from database. It got merged today, so you'll see it in the next nightly.

Reply:
Yep, you're of course right - mixed issues up. Nevertheless, feels incredibly good. ;)